airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/PartFactory.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.file.object_storage
+
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Generates part w/ metadata for a multi-part upload for a given key and file no. parts are
+ * 1-indexed. For convenience, empty parts are tolerated but not counted by the assembler.
+ *
+ * Not thread-safe. It is expected that the parts are generated in order.
+ */
+class PartFactory(
+    val key: String,
+    val fileNumber: Long,
+) {
+    var totalSize: Long = 0
+    private var nextIndex: Int = 0
+    private var finalProduced = false
+
+    fun nextPart(bytes: ByteArray?, isFinal: Boolean = false): Part {
+        if (finalProduced) {
+            throw IllegalStateException("Final part already produced")
+        }
+        finalProduced = isFinal
+
+        totalSize += bytes?.size?.toLong() ?: 0
+        // Only advance the index if the part isn't empty.
+        // This way empty parts can be ignored, but empty final parts
+        // can still convey the final index.
+        if (bytes != null) {
+            nextIndex++ // pre-increment as parts are 1-indexed
+        }
+        return Part(
+            key = key,
+            fileNumber = fileNumber,
+            partIndex = nextIndex,
+            bytes = bytes,
+            isFinal = isFinal
+        )
+    }
+}
+
+/**
+ * Reassembles part metadata into a view of the upload state.
```

Comment:
It's the assembler of a view of the batch right? Naming makes it sound like it assembles parts.

Reply:
Yeah, I couldn't come up with a good name. I added the `Metadata` to try to avoid implying that it was assembling *parts* specifically. Maybe `PartUploadStateView`?

Reply:
or maybe even just `PartBookkeeper`?

Reply:
PartBookkeeper aint bad